rspace/src/main/scala/coop/rchain/rspace/state/RSpaceImporter.scala

```diff
@@ -1,8 +1,169 @@
 import coop.rchain.state.TrieImporter
+import fs2.Stream
+import scodec.Codec
+import scodec.bits.ByteVector
 
 trait RSpaceImporter[F[_]] extends TrieImporter[F] {
   type KeyHash = Blake2b256Hash
+
+  def getHistoryItem(hash: KeyHash): F[Option[ByteVector]]
+}
+
+class StateValidationError(message: String) extends Exception(message) {
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  def this(message: String, cause: Throwable) = {
+    this(message)
+    initCause(cause)
+  }
+}
+
+object RSpaceImporter {
+  def validateStateItems[F[_]: Concurrent, C, P, A, K](
+      historyItems: Seq[(Blake2b256Hash, ByteVector)],
+      dataItems: Seq[(Blake2b256Hash, ByteVector)],
+      startPath: Seq[(Blake2b256Hash, Option[Byte])],
+      chunkSize: Int,
+      skip: Int,
+      getFromHistory: Blake2b256Hash => F[Option[ByteVector]]
+  )(implicit codecC: Codec[C], codecP: Codec[P], codecA: Codec[A], codecK: Codec[K]): F[Unit] = {
+    import cats.instances.list._
+
+    val receivedHistorySize = historyItems.size
+    val receivedDataSize    = dataItems.size
+    val isEnd               = receivedHistorySize < chunkSize
+
+    def raiseError[T](msg: String): F[T] = new StateValidationError(msg).raiseError[F, T]
+    def raiseErrorEx[T](msg: String, cause: Throwable): F[T] =
+      new StateValidationError(msg, cause).raiseError[F, T]
+
+    def decodeTrie(bytes: ByteVector): Trie = Trie.codecTrie.decodeValue(bytes.bits).get
+
+    def decodeData(bytes: ByteVector) =
+      ColdStoreInstances.codecPersistedData.decodeValue(bytes.bits).toEitherThrowable
+
+    // Validate history items size
+    def validateHistorySize[T]: F[Unit] = {
+      val sizeIsValid = receivedHistorySize == chunkSize | isEnd
+      raiseError[T](
+        s"Input size of history items is not valid. Expected chunk size $chunkSize, received $receivedHistorySize."
+      ).whenA(!sizeIsValid)
+    }
+
+    // Validate data items size
+    def validateDataSize[T]: F[Unit] = {
+      val sizeIsValid = receivedDataSize <= chunkSize | isEnd
+      raiseError[T](
+        s"Input size of data items $receivedDataSize is greater then expected chunk size $chunkSize."
+      ).whenA(!sizeIsValid)
+    }
+
+    // Tries decoded from received items
+    // - validate trie hash to match decoded trie hash
+    def tries: F[List[(Blake2b256Hash, Trie)]] = historyItems.toList traverse {
+      case (hash, trieBytes) =>
+        val trie = decodeTrie(trieBytes)
+        // TODO: this is strange that we need to encode Trie again to get the hash.
+        //  - the reason is that hash is not calculated on result of `Trie.codecTrie.encode`
+        //    but on each case of Trie instance (see `Trie.hash`)
+        //  - this adds substantial time for validation e.g. 20k records 450ms (with encoding 2.4sec)
+        // https://github.com/rchain/rchain/blob/4dd216a7/rspace/src/main/scala/coop/rchain/rspace/history/HistoryStore.scala#L25-L26
+        val trieHash = Trie.hash(trie)
```

Comment:
I think you can get the hash by `val trieHash = Blake2b256Hash.create(trieBytes.bits.drop(2L).toByteVector)`
But I tried and it doesn't  turns out to be the same always.

It might related to the transfer situation like below.
```
val b = BitVector(something)
BitVector(b.toByteVector) != b  (sometimes)
```

Reply:
You are right, this is direct manipulation of serialized data but in this situation it can be justified.

I'll try to change that later as part of additional optimizations with more testing. I also noticed comparison of BitVector's is not consistent but conversion to raw bytes shouldn't be a problem.